lib/Dialect/Moore/MooreOps.cpp

```diff
@@ -371,15 +395,116 @@ void NetOp::getAsmResultNames(OpAsmSetValueNameFn setNameFn) {
+    rewriter.eraseOp(uniqueAssignOp);
+    modified = true;
+  }
+
+  // If all users of the net op are reads (besides the unique assignment),
+  // directly replace the reads with the net's assigned value.
+  if (allUsesAreReads && op.getAssignment()) {
+    // If the original net had a name, create an `AssignedVariableOp` as a
+    // replacement. Otherwise substitute the assigned value directly.
+    auto assignedValue = op.getAssignment();
+    if (auto name = op.getNameAttr(); name && !name.empty())
+      assignedValue = rewriter.create<AssignedVariableOp>(
+          op.getLoc(), name, uniqueAssignOp.getSrc());
+
+    // Replace all reads with the new assigned var op and remove the original
+    // net op.
+    for (auto *user : llvm::make_early_inc_range(op->getUsers())) {
+      auto readOp = cast<ReadOp>(user);
+      rewriter.replaceOp(readOp, assignedValue);
+    }
+    rewriter.eraseOp(op);
+    modified = true;
+  }
+
+  return success(modified);
+}
+
 //===----------------------------------------------------------------------===//
-// AssignedVarOp
+// AssignedVariableOp
 //===----------------------------------------------------------------------===//
 
-void AssignedVarOp::getAsmResultNames(OpAsmSetValueNameFn setNameFn) {
+void AssignedVariableOp::getAsmResultNames(OpAsmSetValueNameFn setNameFn) {
   if (getName() && !getName()->empty())
     setNameFn(getResult(), *getName());
 }
 
+LogicalResult AssignedVariableOp::canonicalize(AssignedVariableOp op,
+                                               PatternRewriter &rewriter) {
+  // Eliminate chained variables with the same name.
+  // var(name, var(name, x)) -> var(name, x)
+  if (auto otherOp = op.getInput().getDefiningOp<AssignedVariableOp>()) {
+    if (otherOp.getNameAttr() == op.getNameAttr()) {
+      rewriter.replaceOp(op, otherOp);
+      return success();
+    }
+  }
+
+  // Eliminate variables that alias an input port of the same name.
+  if (auto blockArg = dyn_cast<BlockArgument>(op.getInput())) {
+    if (auto moduleOp =
+            dyn_cast<SVModuleOp>(blockArg.getOwner()->getParentOp())) {
+      auto moduleType = moduleOp.getModuleType();
+      auto portName = moduleType.getInputNameAttr(blockArg.getArgNumber());
+      if (portName == op.getNameAttr()) {
+        rewriter.replaceOp(op, blockArg);
+        return success();
+      }
+    }
+  }
+
+  // Eliminate variables that feed an output port of the same name.
+  for (auto &use : op->getUses()) {
+    auto outputOp = dyn_cast<OutputOp>(use.getOwner());
+    if (!outputOp)
+      continue;
+    auto moduleOp = dyn_cast<SVModuleOp>(outputOp.getParentOp());
+    if (!moduleOp)
+      break;
+    auto moduleType = moduleOp.getModuleType();
+    auto portName = moduleType.getOutputNameAttr(use.getOperandNumber());
+    if (portName == op.getNameAttr()) {
+      rewriter.replaceOp(op, op.getInput());
+      return success();
+    }
+  }
```

Comment:
I have a question: The type of `use` is OpOperand, in your example `%b = moore.assigned_variable %3 : i42`, so I always think `use` == `%3`, and why `use.getOwner()` will return an output op :thinking:? Similarly, I view the annotation regarding `OpOperand.getOperandNumber()`, because `assigned_variable` only has one operand(`input`) , so `use.getOperandNumber()` must get zero, if we tweak the order of `moore.output %b, %w : !moore.i42, !moore.i42` like `moore.output %w, %b : !moore.i42, !moore.i42`. Is this canonicalization also triggered?

Reply:
You're right that `use` is a `OpOperand`, but importantly the loop goes through all `OpOperand`s of all the other ops that use `op`. In the example of `%b = moore.assigned_variable %3 : i42`, `getUses()` does not return the `%3`, but all places in other ops where a `%b` is used. If there's an op like `moore.output %b, %w` around, calling `getUses()` on `%b` will produce the `OutputOp`'s use of `%b`. If you had `moore.output %b, %b`, then `getUses()` would produce two `OpOperand`s, both with `OutputOp` as the owner, but one with operand number 0 and one with operand number 1.

Basically, `OpOperand` is the slot in an operation where a value can be plugged into. Asking for `getUses()` on an op returns you a list of all these slots where the op (or value if you call it on a value) is used.

Reply:
Thanks for the detailed explanation :smiley: !  Looks like I misunderstand `getUses()`.  But now I know.